Source/JavaScriptCore/llint/WebAssembly.asm

```diff
@@ -1024,6 +1071,149 @@ end
+        macro (callee, targetWasmInstance)
+            move callee, ws0
+
+            loadi ArgumentCountIncludingThis + TagOffset[cfr], PC
+
+            # the call might throw (e.g. indirect call with bad signature)
+            btpz targetWasmInstance, .throw
+
+            wgetu(ctx, m_stackOffset, ws1)
+            muli SlotSize, ws1
+
+if ARMv7
+            subp cfr, ws1, ws1
+            move ws1, sp
+else
+            subp cfr, ws1, sp
+end
+
+            storeWasmInstance(targetWasmInstance)
+            reloadMemoryRegistersFromInstance(targetWasmInstance, wa0, wa1)
+
+            wgetu(ctx, m_numberOfCalleeStackArgs, ws1)
+
+if ARM64 or ARM64E
+            leap [sp, ws1, 8], ws1
+            forEachArgumentJSR(macro (offset, gpr1, gpr2)
+                loadpairq CallFrameHeaderSize + 8 + offset[ws1], gpr1, gpr2
+            end)
+elsif JSVALUE64
+            forEachArgumentJSR(macro (offset, gpr)
+                loadq CallFrameHeaderSize + 8 + offset[sp, ws1, 8], gpr
+            end)
+else
+            forEachArgumentJSR(macro (offset, gprMsw, gpLsw)
+                load2ia CallFrameHeaderSize + 8 + offset[sp, ws1, 8], gpLsw, gprMsw
+            end)
+end
+if ARM64 or ARM64E
+            forEachArgumentFPR(macro (offset, fpr1, fpr2)
+                loadpaird CallFrameHeaderSize + 8 + offset[ws1], fpr1, fpr2
+            end)
+else
+            forEachArgumentFPR(macro (offset, fpr)
+                loadd CallFrameHeaderSize + 8 + offset[sp, ws1, 8], fpr
+            end)
+end
+
+            # We need PC to load information from context
+            # and two scratch registers to perform the copy
+            preserveGPRsUsedByTailCall(wa0, wa1)
+            move PC, ws1
+            loadi ArgumentCountIncludingThis + TagOffset[cfr], PC
+
+            # old stack size
+            wgetu(ctx, m_numberOfCallerStackArgs, wa0)
+            muli SlotSize, wa0
+            addi CallFrameHeaderSize + StackAlignment - 1, wa0
+            andi ~StackAlignmentMask, wa0
+
+            # new stack size
+            wgetu(ctx, m_numberOfCalleeStackArgs, wa1)
+            muli SlotSize, wa1
+            addi CallFrameHeaderSize + StackAlignment - 1, wa1
+            andi ~StackAlignmentMask, wa1
+
+            # new stack pointer
+            addp cfr, wa0
+            subp wa1 , wa0
+
+            move ws1, PC
+            preserveReturnAddress(ws1)
+            usePreviousFrame()
+
+if ARMv7
+            addi SlotSize, wa1
+end
+            .copyLoop:
+if ARM64 and not ADDRESS64
+                    subi MachineRegisterSize, wa1
+                    loadq [sp, wa1, 1], ws1
+                    storeq ws1, [wa0, wa1, 1]
+                    btinz wa1, .copyLoop
+else
+                    subp MachineRegisterSize, wa1
+                    loadp [sp, wa1, 1], ws1
+                    storep ws1, [wa0, wa1, 1]
+                    btinz wa1, .copyLoop
+end
+
+            move wa0, ws1
+            restoreGPRsUsedByTailCall(wa0, wa1)
+
```

Comment:
A comment would be useful here. 
 
So at this point, cfr is the caller's fp, and wa0 is the new callee sp. The callee hasn't had a chance to allocate any locals yet, so callee sp == callee fp at this point (and we couldn't know how many it would have anyway because of call_indirect). Our args to callee are stored right above our SP. We have moved our stack args from our stack to their final position now.

My question: haven't we clobbered the first two arguments, stored in wa0 and wa1? I guess this would only show up on ARM64, but then why aren't your tests broken there? :confused_pikachu_face:

Reply:
Will do for the comment.

> My question: haven't we clobbered the first two arguments, stored in wa0 and wa1? I guess this would only show up on ARM64, but then why aren't your tests broken there? :confused_pikachu_face:
The two gprs have been saved on the stack in `preserveGPRsUsedByTailCall`, they are stored at sp+CodeBlock and sp+Callee. I chose these slots because they are only populated later in the function prologue.
We then clobber them when copying, and restore them before updating sp.